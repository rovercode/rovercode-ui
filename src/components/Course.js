import React from 'react';
import { hot } from 'react-hot-loader';
import {
  Accordion,
  AccordionSummary,
  AccordionDetails,
  Grid,
  Typography,
} from '@material-ui/core';
import { ExpandMore } from '@material-ui/icons';
import PropTypes from 'prop-types';

import LessonCard from './LessonCard';

const Course = ({ course, userTier, onLessonClick }) => (
  <Accordion defaultExpanded>
    <AccordionSummary expandIcon={<ExpandMore />}>
      <Typography>
        { course.name }
      </Typography>
    </AccordionSummary>
    <AccordionDetails>
      <Grid container direction="row" spacing={2}>
        {
          course.lessons.sort((l1, l2) => l1.sequence_number - l2.sequence_number)
            .map((lesson) => (
              <Grid item xs={12} md={6} lg={3} key={lesson.id}>
                <LessonCard lesson={lesson} userTier={userTier} onClick={onLessonClick} />
              </Grid>
            ))
        }
      </Grid>
    </AccordionDetails>
  </Accordion>
);

Course.defaultProps = {
  userTier: 1,
};

Course.propTypes = {
  course: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    lessons: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      reference: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
    })).isRequired,
  }).isRequired,
  userTier: PropTypes.number,
  onLessonClick: PropTypes.func.isRequired,
};

export default hot(module)(Course);
