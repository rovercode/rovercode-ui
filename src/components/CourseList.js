import React, { Component } from 'react';
import { hot } from 'react-hot-loader';
import {
  Box,
  Button,
  CircularProgress,
  Container,
  Divider,
  Grid,
  Typography,
  TextField,
  InputAdornment,
  Menu,
  MenuItem,
} from '@material-ui/core';
import { Pagination } from '@material-ui/lab';
import { withStyles } from '@material-ui/core/styles';
import {
  ArrowDownward,
  ArrowUpward,
  Search,
} from '@material-ui/icons';
import { FormattedMessage, injectIntl } from 'react-intl';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import Course from './Course';
import Footer from './Footer';

const styles = (theme) => ({
  mainContainer: {
    marginBottom: theme.spacing(8),
    [theme.breakpoints.up('xs')]: {
      minWidth: theme.breakpoints.values.xs,
    },
    [theme.breakpoints.up('sm')]: {
      minWidth: theme.breakpoints.values.sm,
    },
    [theme.breakpoints.up('md')]: {
      minWidth: theme.breakpoints.values.md,
    },
    [theme.breakpoints.up('lg')]: {
      minWidth: theme.breakpoints.values.lg,
    },
  },
  listContainer: {
    marginTop: theme.spacing(2),
  },
  paginationPaddedBox: {
    display: 'flex',
    justifyContent: 'center',
    marginTop: theme.spacing(3),
    marginBottom: theme.spacing(3),
  },
});

class CourseList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      page: 1,
      searchQuery: null,
      ordering: 'name',
      sortMenuAnchorElement: null,
      programSelected: null,
    };
  }

  componentDidMount() {
    const { fetchCourses } = this.props;

    return fetchCourses();
  }

  update = () => {
    const { fetchCourses } = this.props;
    const {
      page,
      ordering,
      searchQuery,
    } = this.state;

    const params = {
      page,
      ordering,
    };

    if (searchQuery) {
      params.search = searchQuery;
    }

    fetchCourses(params);
  }

  selectProgram = (e) => {
    let program = e.target.parentNode;
    if (e.target.parentNode.parentNode.parentNode.id) {
      program = e.target.parentNode.parentNode.parentNode;
    } else if (e.target.parentNode.parentNode.parentNode.parentNode.parentNode.id) {
      program = e.target.parentNode.parentNode.parentNode.parentNode.parentNode;
    }

    this.setState({
      programSelected: program.id,
    });
  }

  toggleOrdering = (name) => {
    const { ordering } = this.state;

    if (ordering.endsWith(name)) {
      if (ordering.startsWith('-')) {
        return ordering.substr(1);
      }

      return `-${ordering}`;
    }

    return name;
  }

  handleSortClick = (event) => this.setState({
    sortMenuAnchorElement: event.currentTarget,
  })

  handleSortClose = () => this.setState({
    sortMenuAnchorElement: null,
  })

  handlePageChange = (e, page) => this.setState({
    page,
  }, () => this.update())

  handleSearchChange = (event) => this.setState({
    searchQuery: event.target.value,
    page: 1,
  }, () => this.update())

  handleOrderingChange = (e) => this.setState({
    ordering: this.toggleOrdering(e.target.id),
  }, () => this.update())

  render() {
    const {
      classes,
      intl,
      courses,
      user,
    } = this.props;
    const {
      ordering,
      programSelected,
      sortMenuAnchorElement,
    } = this.state;

    const searchPlaceholder = intl.formatMessage({
      id: 'app.course_list.search',
      description: 'Placeholder for search entry',
      defaultMessage: 'Search courses',
    });

    const title = intl.formatMessage({
      id: 'app.course_list.title',
      description: 'Header for all courses',
      defaultMessage: 'Courses',
    });

    const Title = withStyles(() => ({
      root: {
        fontWeight: 'bold',
      },
    }))(Typography);

    const TitleArea = withStyles((theme) => ({
      root: {
        marginBottom: theme.spacing(4),
      },
    }))(Grid);

    return (
      <>
        <Container className={classes.mainContainer}>
          <TitleArea item container direction="row" justify="space-between">
            <Grid item>
              <Title variant="h4">
                { title }
              </Title>
            </Grid>
          </TitleArea>
          <Grid container spacing={2}>
            <Grid item xs={4}>
              <TextField
                variant="outlined"
                size="small"
                className="prompt"
                icon="search"
                placeholder={searchPlaceholder}
                onChange={this.handleSearchChange}
                InputProps={{
                  endAdornment: (
                    <InputAdornment position="end">
                      <Search />
                    </InputAdornment>
                  ),
                }}
              />
            </Grid>
            <Grid item xs={8}>
              <Grid container justify="flex-end" spacing={2}>
                <Grid item>
                  <Button aria-controls="sort-menu" aria-haspopup="true" onClick={this.handleSortClick}>
                    <FormattedMessage
                      id="app.program_collection.sort"
                      description="Button label for sort options"
                      defaultMessage="Sort"
                    />
                  </Button>
                  <Menu
                    id="sort-menu"
                    anchorEl={sortMenuAnchorElement}
                    keepMounted
                    open={Boolean(sortMenuAnchorElement)}
                    onClose={this.handleSortClose}
                    getContentAnchorEl={null}
                    anchorOrigin={{
                      vertical: 'bottom',
                      horizontal: 'center',
                    }}
                    transformOrigin={{
                      vertical: 'top',
                      horizontal: 'center',
                    }}
                  >
                    <MenuItem onClick={this.handleOrderingChange} id="name">
                      <FormattedMessage
                        id="app.program_collection.name"
                        description="Button label to sort by name"
                        defaultMessage="Name"
                      />
                      {
                      ordering === 'name' ? (
                        <>
                          <ArrowDownward />
                        </>
                      ) : (
                        <>
                          <ArrowUpward />
                        </>
                      )
                    }
                    </MenuItem>
                  </Menu>
                </Grid>
              </Grid>
            </Grid>
          </Grid>
          <Grid container direction="column" justify="center" alignItems="stretch">
            {
              courses === null ? (
                <Grid item container direction="row" justify="center">
                  <CircularProgress />
                </Grid>
              ) : (null)
            }
            {
              courses && courses.count === 0 ? (
                <Grid item container direction="row" alignItems="center" justify="center">
                  <Typography variant="h4">
                    <FormattedMessage
                      id="app.course_list.nothing"
                      description="Informs the user that no courses match the filters"
                      defaultMessage="Sorry, no courses match your filters."
                    />
                  </Typography>
                </Grid>
              ) : (
                <Grid item>
                  <Box
                    className={classes.listContainer}
                  >
                    {
                      courses ? courses.results.map((course) => (
                        <Course
                          key={course.id}
                          course={course}
                          onLessonClick={this.selectProgram}
                          userTier={user.tier}
                        />
                      )) : (null)
                    }
                  </Box>
                </Grid>
              )
            }
          </Grid>
          {
          courses && courses.total_pages > 1 ? (
            <Box
              className={classes.paginationPaddedBox}
            >
              <Pagination
                defaultPage={1}
                count={courses.total_pages}
                showFirstButton
                showLastButton
                color="secondary"
                onChange={this.handlePageChange}
              />
            </Box>
          ) : (null)
        }
        </Container>
        <Divider />
        <Footer />
        {
          programSelected ? (
            <Redirect to={{
              pathname: `/mission-control/${programSelected}`,
            }}
            />
          ) : (null)
        }
      </>
    );
  }
}

CourseList.defaultProps = {
  courses: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
};

CourseList.propTypes = {
  classes: PropTypes.shape({
    mainContainer: PropTypes.string.isRequired,
    listContainer: PropTypes.string.isRequired,
    paginationPaddedBox: PropTypes.string.isRequired,
  }).isRequired,
  fetchCourses: PropTypes.func.isRequired,
  courses: PropTypes.shape({
    count: PropTypes.number,
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        lessons: PropTypes.arrayOf(PropTypes.shape({
          id: PropTypes.number.isRequired,
          reference: PropTypes.string.isRequired,
          description: PropTypes.string.isRequired,
        })).isRequired,
      }),
    ),
  }),
  user: PropTypes.shape({
    tier: PropTypes.number,
  }).isRequired,
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
};

export default hot(module)(injectIntl(withStyles(styles)(CourseList)));
