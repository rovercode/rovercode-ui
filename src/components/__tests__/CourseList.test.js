import React from 'react';
import { MemoryRouter } from 'react-router';
import { Card, CircularProgress } from '@material-ui/core';
import { Redirect } from 'react-router-dom';
import Course from '../Course';
import CourseList from '../CourseList';

describe('The CourseList component', () => {
  const fetchCourses = jest.fn().mockResolvedValue();
  const courses = {
    count: 2,
    next: null,
    previous: null,
    total_pages: 2,
    results: [{
      id: 1,
      name: 'Test',
      description: 'This is a test',
      lessons: [{
        id: 1,
        reference: 'Lesson1',
        description: 'The first lesson',
        sequence_number: 1,
        active_bd: 1,
        active_bd_owned: false,
        state: {
          progress: 'IN_PROGRESS',
        },
      }, {
        id: 2,
        reference: 'Lesson2',
        description: 'The second lesson',
        sequence_number: 1,
        active_bd: 2,
        active_bd_owned: true,
        state: {
          progress: 'AVAILABLE',
        },
      }],
    }],
  };
  const user = {
    tier: 1,
  };

  beforeEach(() => {
    fetchCourses.mockReset();
  });

  test('renders on the page with no errors', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches courses on mount', async () => {
    await mountWithIntl(
      <MemoryRouter>
        <CourseList
          fetchCourses={fetchCourses}
          user={user}
        />
      </MemoryRouter>,
    );
    expect(fetchCourses.mock.calls.length).toBe(1);
  });

  test('shows the correct number of courses for the user', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    expect(wrapper.find(Course).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).exists()).toBe(false);
  });

  test('shows loading when courses fetching', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        fetchCourses={fetchCourses}
        user={user}
        courses={null}
      />,
    ).dive().dive().dive();

    expect(wrapper.find(Course).exists()).toBe(false);
    expect(wrapper.find(CircularProgress).exists()).toBe(true);
    expect(wrapper.find(CircularProgress).length).toBe(1);
  });

  test('shows message when no results', () => {
    const empty = {
      count: 0,
      next: null,
      previous: null,
      total_pages: 1,
      results: [],
    };
    const wrapper = shallowWithIntl(
      <CourseList
        fetchCourses={fetchCourses}
        user={user}
        courses={empty}
      />,
    ).dive().dive().dive();

    expect(wrapper.find(Card).exists()).toBe(false);
    expect(wrapper.find('WithStyles(ForwardRef(Typography))').children().prop('defaultMessage')).toBe('Sorry, no courses match your filters.');
  });

  test('fetches courses after page change', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.instance().handlePageChange(null, 2);

    expect(fetchCourses).toHaveBeenCalledWith({
      ordering: 'name',
      page: 2,
    });
  });

  test('fetches courses after search change', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.instance().handleSearchChange({
      target: {
        value: 'abc',
      },
      page: 1,
    });

    expect(fetchCourses).toHaveBeenCalledWith({
      ordering: 'name',
      page: 1,
      search: 'abc',
    });
  });

  test('fetches courses after ordering change', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.instance().handleOrderingChange({
      target: {
        id: 'name',
      },
    });

    expect(fetchCourses).toHaveBeenCalledWith({
      ordering: '-name',
      page: 1,
    });

    wrapper.instance().handleOrderingChange({
      target: {
        id: 'name',
      },
    });

    expect(fetchCourses).toHaveBeenCalledWith({
      ordering: 'name',
      page: 1,
    });

    wrapper.instance().handleOrderingChange({
      target: {
        id: 'completion',
      },
    });

    expect(fetchCourses).toHaveBeenCalledWith({
      ordering: 'completion',
      page: 1,
    });
  });

  test('anchors the menu', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.instance().handleSortClick({ currentTarget: 'element' });

    expect(wrapper.state('sortMenuAnchorElement')).toBe('element');

    wrapper.instance().handleSortClose();

    expect(wrapper.state('sortMenuAnchorElement')).toBeNull();
  });

  test('redirects to mission control when program loads', () => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.setState({
      programSelected: 1,
    });

    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).at(0).prop('to')).toEqual({
      pathname: '/mission-control/1',
    });
  });

  test('selects a program', (done) => {
    const wrapper = shallowWithIntl(
      <CourseList
        courses={courses}
        user={user}
        fetchCourses={fetchCourses}
      />,
    ).dive().dive().dive();

    wrapper.instance().selectProgram({
      target: {
        parentNode: {
          parentNode: {
            parentNode: {
              id: undefined,
              parentNode: {
                parentNode: {
                  id: 33,
                  dataset: {
                    owned: 'false',
                  },
                },
              },
            },
          },
        },
      },
    });
    expect(wrapper.state('programSelected')).toBe(33);

    wrapper.setState({
      programLoaded: false,
    });

    wrapper.instance().selectProgram({
      target: {
        parentNode: {
          parentNode: {
            parentNode: {
              id: 55,
              dataset: {
                owned: 'true',
              },
            },
          },
        },
      },
    });
    expect(wrapper.state('programSelected')).toBe(55);

    wrapper.setState({
      programLoaded: false,
    });

    wrapper.instance().selectProgram({
      target: {
        parentNode: {
          id: 77,
          dataset: {
            owned: 'false',
          },
          parentNode: {
            parentNode: {
              parentNode: {
                parentNode: {
                },
              },
            },
          },
        },
      },
    });
    expect(wrapper.state('programSelected')).toBe(77);
    done();
  });
});
