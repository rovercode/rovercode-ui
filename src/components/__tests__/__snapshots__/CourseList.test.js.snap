// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The CourseList component renders on the page with no errors 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Container))
    className="CourseList-mainContainer-1"
  >
    <WithStyles(WithStyles(ForwardRef(Grid)))
      container={true}
      direction="row"
      item={true}
      justify="space-between"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(WithStyles(ForwardRef(Typography)))
          variant="h4"
        >
          Courses
        </WithStyles(WithStyles(ForwardRef(Typography)))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(WithStyles(ForwardRef(Grid)))>
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={4}
      >
        <WithStyles(ForwardRef(TextField))
          InputProps={
            Object {
              "endAdornment": <WithStyles(ForwardRef(InputAdornment))
                position="end"
              >
                <Memo(SearchIcon) />
              </WithStyles(ForwardRef(InputAdornment))>,
            }
          }
          className="prompt"
          icon="search"
          onChange={[Function]}
          placeholder="Search courses"
          size="small"
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={8}
      >
        <WithStyles(ForwardRef(Grid))
          container={true}
          justify="flex-end"
          spacing={2}
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
          >
            <WithStyles(ForwardRef(Button))
              aria-controls="sort-menu"
              aria-haspopup="true"
              onClick={[Function]}
            >
              <FormattedMessage
                defaultMessage="Sort"
                id="app.program_collection.sort"
              />
            </WithStyles(ForwardRef(Button))>
            <WithStyles(ForwardRef(Menu))
              anchorEl={null}
              anchorOrigin={
                Object {
                  "horizontal": "center",
                  "vertical": "bottom",
                }
              }
              getContentAnchorEl={null}
              id="sort-menu"
              keepMounted={true}
              onClose={[Function]}
              open={false}
              transformOrigin={
                Object {
                  "horizontal": "center",
                  "vertical": "top",
                }
              }
            >
              <WithStyles(ForwardRef(MenuItem))
                id="name"
                onClick={[Function]}
              >
                <FormattedMessage
                  defaultMessage="Name"
                  id="app.program_collection.name"
                />
                <ArrowDownwardIcon />
              </WithStyles(ForwardRef(MenuItem))>
            </WithStyles(ForwardRef(Menu))>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      alignItems="stretch"
      container={true}
      direction="column"
      justify="center"
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <Styled(MuiBox)
          className="CourseList-listContainer-2"
        >
          <Course
            course={
              Object {
                "description": "This is a test",
                "id": 1,
                "lessons": Array [
                  Object {
                    "active_bd": 1,
                    "active_bd_owned": false,
                    "description": "The first lesson",
                    "id": 1,
                    "reference": "Lesson1",
                    "sequence_number": 1,
                    "state": Object {
                      "progress": "IN_PROGRESS",
                    },
                  },
                  Object {
                    "active_bd": 2,
                    "active_bd_owned": true,
                    "description": "The second lesson",
                    "id": 2,
                    "reference": "Lesson2",
                    "sequence_number": 1,
                    "state": Object {
                      "progress": "AVAILABLE",
                    },
                  },
                ],
                "name": "Test",
              }
            }
            key="1"
            onLessonClick={[Function]}
            userTier={1}
          />
        </Styled(MuiBox)>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
    <Styled(MuiBox)
      className="CourseList-paginationPaddedBox-3"
    >
      <WithStyles(ForwardRef(Pagination))
        color="secondary"
        count={2}
        defaultPage={1}
        onChange={[Function]}
        showFirstButton={true}
        showLastButton={true}
      />
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Divider)) />
  <Footer />
</Fragment>
`;
